### PR TITLE
finetune notebook added

### PR DESCRIPTION
**Pull request by awsaf49**
Hi, I have created a notebook to showcase **fine-tuning** using **ConvNext-TF** on **Flower Classification** Dataset. It supports both **TPU** and **GPU**. 
Let me know if it checks the requirements.

### CODE REVIEW

**Comment by sayakpaul** [comment]
Thanks for the PR! I appreciate it :) 

It might be a good idea to follow [this notebook](https://github.com/sayakpaul/ViT-jax2tf/blob/main/fine_tune.ipynb) to get an idea of how I structure notebooks. 

For your case, here are some comments:

* The imports need to be organized properly. Refer to the above-mentioned notebook. 
* Tensorflow -> TensorFlow.
* Remove commented code (e.g., `# strategy = tf.distribute.MirroredStrategy() # for GPU or multi-GPU machines`).
* Group together similar code blocks and add a sufficient amount of spacing in between them. The cells just under "Configuration" could be made more tidy following this approach. 
* Format the code cells with [Black](https://github.com/drillan/jupyter-black). 
* Closely follow the image normalization operation. You are simply scaling the images to the range of [0, 1] whereas ConvNeXt models take ImageNet-1k stat normalized inputs. This needs to be followed especially since we are fine-tuning the models. Refer to [this notebook](https://github.com/sayakpaul/ConvNeXt-TF/blob/main/i1k_eval/eval.ipynb) to see how it should be done. 
* We are missing out on the preprocessing utilities. While I understand not all of them might be possible to implement here I'd expect some at least e.g., random cropping, random flipping, etc. 
* You're unfreezing the model all at once but using a way higher LR. This will break the pre-trained weights quickly and the model will not be able to adapt to the downstream objective. So, including an LR schedule that respects this phenomenon would be helpful. You can refer to the above-mentioned notebook for one. 
* Maintain consistency in the comments e.g., their casing, using full stops, etc. 

Let me know if it's unclear.

**Comment by awsaf49** [comment]
Thanks for the review, I'll try to modify the code according to the comments ...

**Comment by awsaf49** [comment]
@sayakpaul I tried to add all the above-mentioned factors. Please have a look
Regarding preprocess,
* As the dataset already comes with `224x224` image_size, **RandomCrop** may not be that effective.
* Still I tried to add **RandomCrop** but it somewhat worsen the result. As the cropped images sometimes miss the part with flowers it is barely recognizable by me.  Hence, didn't add it. The cropping dimension could be tuned to fix this issue.

**Comment by sayakpaul** [comment]
Simply amazing work! Thank you so much! Just one comment. 

Just adjust the spacing like so:

```py
# Model
IMAGE_SIZE = [224, 224]
MODEL_PATH = "gs://convnext/saved_models/convnext_base_1k_224"

# TPU
if tpu:
    BATCH_SIZE = (
        16 * strategy.num_replicas_in_sync
    )  # a TPU has 8 cores so this will be 128
else:
    BATCH_SIZE = 32  # on Colab/GPU, a higher batch size may throw(OOM)

# Dataset
CLASSES = [
    "dandelion",
    "daisy",
    "tulips",
    "sunflowers",
    "roses",
]  # don't change the order

# Other constants
MEAN = tf.constant([0.485 * 255, 0.456 * 255, 0.406 * 255])  # imagenet mean
STD = tf.constant([0.229 * 255, 0.224 * 255, 0.225 * 255])  # imagenet std
AUTO = tf.data.AUTOTUNE
```

With this change, we are good to go.

**Comment by sayakpaul** [comment]
Thanks for the update, @awsaf49. I think we can merge.

**Comment by awsaf49** [comment]
@sayakpaul Thanks.
Btw I tried freezing the `LayerNormalization` layer and the result kinda drops from `0.94+` to `0.92+` with early overfitting. I can't seem to remember the paper but their result showed "Full Fine Tune most likely to score better if one can tune them correctly"

**Comment by awsaf49** [comment]
> @sayakpaul Thanks. Btw I tried freezing the `LayerNormalization` layer and the result kinda drops from `0.94+` to `0.92+` with early overfitting. I can't seem to remember the paper but their result showed "Full Fine Tune most likely to score better if one can tune them correctly"

Here's the comparison with and without freezing `LayerNormalization`:
Without freeze:
image

With freeze:
image

**Comment by sayakpaul** [comment]
Thanks for being rigorous. For Batch Norm though, it's always advisable to freeze it if we are in the fine-tuning regime. I guess that does not apply to Layer Norm.

**Comment by awsaf49** [comment]
Yes, you are right, freezing the `BatchNormlization` layer often results in a better score.
